### ISSUE TITLE
"dangling symbolic link" flakes after upgrading to Bazel 7

### ISSUE THREAD

**Issue by dbolduc** [bazelbuild/bazel#20886]
### Description of the bug:

https://github.com/googleapis/google-cloud-cpp/issues/13444

After upgrading to Bazel 7, we have started seeing transient failures in our CI. These have all been from [`io_opentelemetry_cpp`](https://github.com/open-telemetry/opentelemetry-cpp).

```
ERROR: /h/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/external/io_opentelemetry_cpp/sdk/src/common/BUILD:6:11: output 'external/io_opentelemetry_cpp/sdk/src/common/_virtual_includes/random/src/common/random.h' is a dangling symbolic link
ERROR: /h/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/external/io_opentelemetry_cpp/sdk/src/trace/BUILD:6:11: output 'external/io_opentelemetry_cpp/sdk/src/trace/_virtual_includes/trace/src/trace/span.h' is a dangling symbolic link
...etc...
ERROR: /h/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/external/io_opentelemetry_cpp/api/BUILD:13:11: Symlinking virtual headers for api failed: not all outputs were created or valid
ERROR: /h/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/external/io_opentelemetry_cpp/sdk/BUILD:6:11: Symlinking virtual headers for headers failed: not all outputs were created or valid
```

My naive guess is that it has something to do with how that repo uses `include_prefix`: https://github.com/open-telemetry/opentelemetry-cpp/blob/c4f39f2be8109fd1a3e047677c09cf47954b92db/sdk/src/trace/BUILD#L10

### Which category does this issue belong to?

External Dependency

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Not sure, but I can supply more logs and test solutions (within reason).

### Which operating system are you running Bazel on?

Linux

### What is the output of `bazel info release`?

release 7.0.0

### Is this a regression? If yes, please try to identify the Bazel commit where the bug was introduced.

It probably has to do with "build without the bytes"

### Have you found anything relevant by searching the web?

https://github.com/bazelbuild/bazel/issues/19143 seems like a similar issue.

**Comment by tjgq** [comment]
Can you share the Bazel flags you're using?

We've seen a few symlink-related issues due to the `--incompatible_sandbox_hermetic_tmp` flag flip. Does setting `--noincompatible_sandbox_hermetic_tmp` change anything?

**Comment by sgowroji** [comment]
Hi @dbolduc, Can you please take a look on the above [comment](https://github.com/bazelbuild/bazel/issues/20886#issuecomment-1891653650).

**Comment by alevenberg** [comment]
```
ERROR: /h/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/external/io_opentelemetry_cpp/api/BUILD:13:11: output 'external/io_opentelemetry_cpp/api/_virtual_includes/api/opentelemetry/version.h' is a dangling symbolic link
ERROR: /h/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/external/io_opentelemetry_cpp/sdk/src/trace/BUILD:6:11: output 'external/io_opentelemetry_cpp/sdk/src/trace/_virtual_includes/trace/src/trace/span.h' is a dangling symbolic link
ERROR: /h/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/external/io_opentelemetry_cpp/sdk/src/trace/BUILD:6:11: Symlinking virtual headers for trace failed: not all outputs were created or valid
ERROR: /h/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/external/io_opentelemetry_cpp/api/BUILD:13:11: Symlinking virtual headers for api failed: not all outputs were created or valid
ERROR: /h/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/external/io_opentelemetry_cpp/sdk/src/common/BUILD:6:11: output 'external/io_opentelemetry_cpp/sdk/src/common/_virtual_includes/random/src/common/random.h' is a dangling symbolic link
ERROR: /h/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/external/io_opentelemetry_cpp/sdk/src/common/BUILD:6:11: Symlinking virtual headers for random failed: not all outputs were created or valid
Analyzing: 859 targets (0 packages loaded, 12520 targets configured)`
```

coverage-ci https://console.cloud.google.com/cloud-build/builds;region=us-east1/62bf2763-5a18-4983-87fb-47254671c10c?project=936212892354

**Comment by dbolduc** [comment]
> Not sure, but I can supply more logs and test solutions (within reason).

I lied.

> Does setting `--noincompatible_sandbox_hermetic_tmp` change anything?

I reverted our builds to use Bazel 6.4.0 before testing this. :shrug:

**Comment by moroten** [comment]
https://github.com/bazelbuild/bazel/issues/20408#issuecomment-1863714833 reports that this is still a problem in Bazel 7.0.0. Even if the PR #19739 (fixing #19143 in 6.4) mentions that it is working at HEAD at the time, before 7.0 was branched, 7.0.0 is still broken (transiently).

**Comment by Gormo** [comment]
This occurs quite frequently in our environment and even more frequently when enabling bzl-module support.
We are using remote execution and build-without-the-bytes but applying the fix in #19739 does not seem to affect anything.

I have tried to recreate this in a minimal sandbox without success but my guess is that it does not seem related to remote-execution since it appears on some targets like this:

cc_library(
  name = "foo",
  **hdrs = glob(["**/*.h"]),**
  srcs = glob(["**/*.c"]),
  **strip_include_prefix = "include",**
)

Where the strip_include_prefix triggers the symlink actions to occur but it fails locally even though the target files must be available locally since they were found by the glob in the repository.

This is a major issue for us and a stopper for us in order to proceed with the migration to Bazel modules.

**Comment by uhlajs** [comment]
Same when using [rules_foreign_cc](https://bazelbuild.github.io/rules_foreign_cc/main/cmake.html). But using `--noincompatible_sandbox_hermetic_tmp` as @tjgq  suggests [here](https://github.com/bazelbuild/bazel/issues/20886#issuecomment-1891653650) fix the issue.

**Comment by tjgq** [comment]
@uhlajs Is this with 7.0.0, or 7.0.2? The latter contains some fixes for `--incompatible_sandbox_hermetic_tmp`.

**Comment by fmeum** [comment]
@uhlajs `rules_foreign_cc`, when used with `cmake`, has some actual issues with hermetic tmp that I think come from `cmake install` not following symlinks. I would recommend filing an issue in the `rules_foreign_cc` if that's what you are running into. Feel free to mention me on it.

**Comment by tjgq** [comment]
Thanks for rerouting this, @fmeum. Feel free to ping or reopen if it turns out to be a Bazel issue and not just a rules_foreign_cc one.

**Comment by Gormo** [comment]
@tjgq the dangling symbolic link issues we were seeing were not related to rules_foreign_cc and are definitely a Bazel issue. I don't think this issue should be closed since this needs to be fixed.

**Comment by moroten** [comment]
@Gormo did not use cmake or `rules_foreign_cc` and he was using 7.0.2 (see his comment above). It is therefore also an issue in Bazel itself when using a native `cc_library`.

**Comment by tjgq** [comment]
Sorry, I misread the thread.

**Comment by moroten** [comment]
@tjgq no problems. Thank you for reopening it.

**Comment by fmeum** [comment]
@Gormo Since it appears to be difficult to produce a reproducer for this, could you perhaps try to bisect this down to the breaking Bazel commit using Bazelisk's `--bisect` feature? Does this still reproduce with hermetic tmp disabled?

**Comment by tjgq** [comment]
@Gormo Before you attempt a bisect, can you check whether `--noexperimental_merged_skyframe_analysis_execution` makes the issue go away? I remembered that Skymeld makes some changes to when and how symlinks to source files are created, and I'm wondering if that could be the root cause.

**Comment by Gormo** [comment]
@tjgq I have triggered the ci pipelines many times now with skymeld disabled and I can no longer see the flaking symlinks. Most likely your theory is correct and skymeld is what's causing the issues!

**Comment by fmeum** [comment]
@bazel-io fork 7 1.0

**Comment by jfirebaugh** [comment]
We (Figma) are also experiencing this issue when attempting a Bazel 7 upgrade. The error looks like:

> ERROR: /home/circleci/.cache/bazel/_bazel_circleci/a6e8cb0ce75ceb1ea1faa0cca0b0ae0d/external/cxx.rs/BUILD:39:11: output 'external/cxx.rs/_virtual_includes/core/rust/cxx.h' is a dangling symbolic link
> ERROR: /home/circleci/.cache/bazel/_bazel_circleci/a6e8cb0ce75ceb1ea1faa0cca0b0ae0d/external/cxx.rs/BUILD:39:11: Symlinking virtual headers for core failed: not all outputs were created or valid

**Comment by Gormo** [comment]
@tjgq just looked at the result of the additional tests that I triggered to run during the night and unfortunately the dangling symlink did occur once even without skymeld. Will try to bisect today!

**Comment by Gormo** [comment]
Something I did notice is that the failure rate seems to be potentially lower with skymeld disabled.

**Comment by tyler-french** [comment]
I am also getting this error only in Bazel 7 from the following `genrule`:

```
genrule(
    name = "copy_files",
    srcs = [
        "@some_repo//:files",
    ],
    outs = [
        "some_file1.json",
        "some_file2.json",
    ],
    cmd = "cp -r $(locations @some_repo//:files) $(RULEDIR)",
)
```

```
ERROR: declared output 'src/...some_file1.json' is a dangling symbolic link
...
ERROR: ...: Executing genrule //src/...:copy_files failed: not all outputs were created or valid
```

**Comment by tjgq** [comment]
@tyler-french Is `@some_repo//:files` a `filegroup` or something else? Are you using Skymeld? Are you building with a disk and/or a remote cache?

**Comment by tyler-french** [comment]
> @tyler-french Is `@some_repo//:files` a `filegroup` or something else? Are you using Skymeld? Are you building with a disk and/or a remote cache?

A `filegroup` yes, disk-cache, not remote cache, no skymeld. Should I try `--noincompatible_sandbox_hermetic_tmp`?

```
    http_archive(
        name = "some_repo",
        ...
        build_file_content = """
filegroup(
    name = "files",
    srcs = glob(["**/*.json"]),
    visibility = ["//visibility:public"]
)
        """,
    )
```

Bazel Flags:
```
 --noexperimental_merged_skyframe_analysis_execution --incompatible_generated_protos_in_virtual_imports --incompatible_strict_action_env --proto_compiler //rules/proto:protoc --disk_cache=~/.bazelcache --experimental_profile_include_target_label --noslim_profile --protocopt=--include_source_info --@io_bazel_rules_docker//transitions:enable=false --experimental_output_directory_naming_scheme=diff_against_dynamic_baseline --experimental_downloader_config=third_party/bazel_downloader.cfg --incompatible_remote_use_new_exit_code_for_lost_inputs --incompatible_enable_cc_toolchain_resolution --extra_toolchains=@zig_sdk//toolchain:linux_amd64_gnu.2.19 --extra_toolchains=@zig_sdk//toolchain:linux_arm64_gnu.2.28 --experimental_reuse_sandbox_directories --@io_bazel_rules_go//go/config:gc_goopts=-d=loopvar=1
```

**Comment by tyler-french** [comment]
Was able to get this working using `--noincompatible_sandbox_hermetic_tmp` cc @tjgq. However, we would probably like to remove this flag if possible.

**Comment by tjgq** [comment]
Here's an even more minimal, 100% reproducible repro, with no flags required other than `--noenable_bzlmod` (just so I can use WORKSPACE):

repo.bzl
```
def _impl(rctx):
  rctx.file("file.txt", "hello")
  rctx.file("BUILD", """exports_files(["file.txt"])""")
```

WORSKPACE
```
load(":repo.bzl", "repo")
repo(name= "repo")
```

BUILD
```
genrule(
  name = "gen",
  outs = ["file.txt"],
  srcs = ["@repo//:file.txt"],
  cmd = "cp -r $(location @repo//:file.txt) $(RULEDIR)",
)
```

Result:
```
ERROR: /usr/local/google/home/tjgq/proj/dangling-symlink-repro/BUILD:1:8: declared output 'file.txt' is a dangling symbolic link
ERROR: /usr/local/google/home/tjgq/proj/dangling-symlink-repro/BUILD:1:8: Executing genrule //:gen failed: not all outputs were created or valid
```

This would seemingly imply that copying a source file into the output tree is thoroughly broken with `--incompatible_hermetic_sandbox_tmp`, which is pretty bad :(

When I use the suggested command to spawn a shell inside the sandbox, I observe the following:
* `pwd` is `/tmp/bazel-working-directory/__main__`
* `external/repo/file.txt` is a symlink to `/tmp/bazel-source-roots/1/file.txt` (which exists)
* `cp -r external/repo/file.txt bazel-out/k8-fastbuild/bin` (which should be the same as the genrule command line) successfully creates a symlink at `bazel-out/k8-fastbuild/bin/file.txt`, which also points to `/tmp/bazel-source-roots/1/file.txt`

However, if I run the action with `--sandbox_debug` and then inspect the sandbox contents, `bazel-out/k8-fastbuild/bin/file.txt` doesn't exist at all, and `external/repo/file.txt` is a dangling symlink to `/tmp/bazel-source-roots/1/file.txt` (the latter isn't surprising since the namespace mounts no longer exist, but I don't understand how the former could be).

@fmeum @lberki maybe what I wrote above rings a bell? I will return to this at some point, but I have another fire to fight at the moment so I can't give it my full attention.

**Comment by tjgq** [comment]
An important feature of this repro (that initially eluded me) is that `cp -r symlink destination` copies the symlink itself, while `cp symlink destination` without the `-r` copies the contents of the file the symlink points to. If I remove the `-r`, the action succeeds, even though `bazel-out/k8-fastbuild/bin/file.txt` is still missing from the sandbox when inspected after Bazel exists (at this point I think I just don't understand how the sandbox works).

**Comment by sbjorkle-qualcomm** [comment]
We do see this frequently even with "--noincompatible_sandbox_hermetic_tmp" when bazel modules is enabled.
Sure, there may be issues with the hermetic tmp handling but that may not be the only root-cause. 
Disabling skymeld seems to reduce the frequency of the failures significantly so some kind of internal race-condition between threads would be my guess.

**Comment by tjgq** [comment]
Yeah, the issue described in https://github.com/bazelbuild/bazel/issues/20886#issuecomment-1928038417 seems to be of a different nature (100% deterministic, Skymeld independent, fixed by disabling the hermetic sandbox). Let's move that one to https://github.com/bazelbuild/bazel/issues/21215.

There is something else we haven't narrowed down yet (non-deterministic, Skymeld-independent, not fixed by disabling the hermetic sandbox). The common thread among the other reports is virtual includes, so I'm wondering if there's something special about the symlinks created by the "Symlinking virtual headers" action, which passes a mysterious (to me) [`use_exec_root_for_source` parameter](https://cs.opensource.google/bazel/bazel/+/master:src/main/starlark/builtins_bzl/common/cc/cc_compilation_helper.bzl;l=124?q=use_exec_root_for_source&ss=bazel) to `ctx.actions.symlink`.

**Comment by tjgq** [comment]
The `use_exec_root_for_source` is a no-op for Bazel (it only exists for the benefit of Blaze), so I think that's a red herring?

@Gormo, any luck in trying to bisect this?

I have thus far failed to reproduce this, and I don't have a clue where to look, so until someone is able to provide a repro (it can be nondeterministic, as long as it's able to produce the error if run enough times) I think it's going to be difficult to make progress on this.

**Comment by Gormo** [comment]
@tjgq unfortunately bisecting using bazelisk was a no-go since it doesn't seem to support undeterministic behaviours... I will check if I can create some kind of reproducible minimal example tomorrow if I have time.

**Comment by tjgq** [comment]
Yeah, either that or you'd have to write your own bisection script that does N trials per iteration. (IMO it would be a reasonable feature request, but changing Bazelisk is probably more work, considering that `git bisect` exists.)

**Comment by fmeum** [comment]
@tjgq It's only a noop since https://github.com/bazelbuild/bazel/pull/20540, which is not in 7.0.2.

**Comment by Gormo** [comment]
An update from our side:
We have tried to recreate this issue in a controlled manor without success.
Observations:
* Occurs more frequently with bazel-modules enabled
* Occurs more frequently on our CI machines which uses vmware with virtual disks.
* Setting --noincompatible_sandbox_hermetic_tmp did not decrease the frequency of the flakes.
* Disabling skymeld decreases the failure rate significantly but does not the remove the flakes completely.

Current theory is that it could be related to io-access and some kind of internal race-condition between different threads and sometimes io-access is delayed on disks with heavy load which triggers this issue.

We tried to recreate this locally by using a disk-loader for simulating high io-load but unfortunately without success.

**Comment by fmeum** [comment]
@Gormo Could you check whether the issue still occurs remotely with 48ea3d2a56d4c940c61200245473a51b78d6f4df (a commit in the 7.1.0 branch that should be available with Bazelisk)? That commit changes how the C++ header symlinks are created in Bazel.

**Comment by Gormo** [comment]
@fmeum I have now tried, cherry-picking 48ea3d2 on top on 7.0.2 but that didn't really affect anything since we get the errors also on ctx.actions.symlink() generated actions.
Then I tried adding a patch to always use "use_exec_root_for_source".

But that didn't remove any flakiness either.

```
diff --git a/src/main/java/com/google/devtools/build/lib/analysis/starlark/StarlarkActionFactory.java b/src/main/java/com/google/devtools/build/lib/analysis/starlark/StarlarkActionFactory.java
index 44ff502fdf..3d037752bb 100644
--- a/src/main/java/com/google/devtools/build/lib/analysis/starlark/StarlarkActionFactory.java
+++ b/src/main/java/com/google/devtools/build/lib/analysis/starlark/StarlarkActionFactory.java
@@ -291,9 +291,7 @@ public class StarlarkActionFactory implements StarlarkActionFactoryApi {
     if (useExecRootForSourceObject != Starlark.UNBOUND) {
       BuiltinRestriction.failIfCalledOutsideAllowlist(thread, PRIVATE_STARLARKIFICATION_ALLOWLIST);
     }
-    boolean useExecRootForSource =
-        !Starlark.UNBOUND.equals(useExecRootForSourceObject)
-            && (Boolean) useExecRootForSourceObject;
+    boolean useExecRootForSource = true;
 
     RuleContext ruleContext = getRuleContext();
```

**Comment by lberki** [comment]
@Wyverald FYI if you haven't already seen this thread

**Comment by lberki** [comment]
@Gormo can you check what the paths look like about which Bazel complains that they are dangling symlinks? I'm curious if they stay dangling symlinks at the end of the build and if so, how exactly they are dangling, i.e. in what step of the symlink resolution does the "file not found" error occur? (e.g. simply the target doesn't exist even though the directory that contains it does? Does the symlink point to a file under a directory that should exist, but it doesn't? Something more complicated?)

**Comment by lberki** [comment]
If it still fails without `--spawn_strategy=standalone`, that exonerates `--incompatible_sandbox_hermetic_tmp`.

Can you check which directory exists and which one does not in the ancestors of `/Top-bazel/478b2cbff2254079381e27d1a245fab2/execroot/_main/bazel-out/aarch64_qnx-opt-ST-d3c7f1d3749b/bin/external/Foo/Configuration/Bazel/Bar/Platform/Foo_proxy_library_stripped.a` and if all of them do,where that symlink points to and what directory among the ancestors of the target exists?

**Comment by Gormo** [comment]
This error also occurs on Windows:
```
09:55:02 ERROR: E:/bazel/ypqoi5y7/external/extrepo/BUILD.bazel:5:11: output 'external/extrepo/_virtual_includes/Foo/foo.h' is a dangling symbolic link
09:55:03 ERROR: E:/bazel/ypqoi5y7/external/extrepo/BUILD.bazel:5:11: Symlinking virtual headers for Foo failed: not all outputs were created or valid
```

Looking at the disk, I can see.
```
myuser@computer MINGW64 /e/folder/project
$ stat bazel-out/k8-fastbuild/bin/external/extrepo/_virtual_includes/Foo/foo.h
  File: bazel-out/k8-fastbuild/bin/external/extrepo/_virtual_includes/Foo/foo.h -> /e/bazel/ypqoi5y7/execroot/AT/external/extrepo/include/foo.h
  Size: 87              Blocks: 0          IO Block: 65536  symbolic link
Device: 769dcb8ch/1990052748d   Inode: 47850746041798830  Links: 1
Access: (0777/lrwxrwxrwx)  Uid: (297610/myuser)   Gid: (297121/ UNKNOWN)
Access: 2024-02-22 09:55:02.963725500 +0100
Modify: 2024-02-22 09:55:02.963725500 +0100
Change: 2024-02-22 09:55:02.963725500 +0100
Birth: 2024-02-22 09:55:02.963725500 +0100

myuser@computer MINGW64 /e/folder/project
$ stat /e/bazel/ypqoi5y7/execroot/AT/external/extrepo/include/foo.h
  File: /e/bazel/ypqoi5y7/execroot/AT/external/extrepo/include/foo.h
  Size: 231955          Blocks: 228        IO Block: 65536  regular file
Device: 769dcb8ch/1990052748d   Inode: 26740122787614710  Links: 1
Access: (0644/-rw-r--r--)  Uid: (297610/myuser)   Gid: (297121/ UNKNOWN)
Access: 2024-02-19 11:48:02.498995600 +0100
Modify: 2024-02-19 11:48:02.498995600 +0100
Change: 2024-02-19 11:48:02.498995600 +0100
Birth: 2024-02-19 11:48:02.495996800 +0100
```

The build event streaming posted the event after the symlink was created.
```
{
  "eventTime": {
    "seconds": "1708592102",  // 2024-02-22 09:55:02
    "nanos": 970000000
  },
  "buildEvent": {
    "id": {
      "actionCompleted": {
        "primaryOutput": "bazel-out/k8-fastbuild/bin/external/extrepo/_virtual_includes/Foo/foo.h",
        "label": "@@extrepo//:Foo",
        "configuration": {
          "id": "6c01f9acdccc727d9bd9c32e7940d9d1ef9fff9fc411e9c028468eb14fcbb30d"
        }
      }
    },
    "action": {
      "exitCode": 1,
      "label": "@@extrepo//:Foo",
      "configuration": {
        "id": "6c01f9acdccc727d9bd9c32e7940d9d1ef9fff9fc411e9c028468eb14fcbb30d"
      },
      "type": "Symlink",
      "failureDetail": {
        "message": "not all outputs were created or valid",
        "execution": {
          "code": "ACTION_OUTPUTS_NOT_CREATED"
        }
      }
    }
  },
}
```

**Comment by freeformstu** [comment]
> @Gormo Could you check whether the issue still occurs with [48ea3d2](https://github.com/bazelbuild/bazel/commit/48ea3d2a56d4c940c61200245473a51b78d6f4df) (a commit in the 7.1.0 branch that should be available with Bazelisk)? That commit changes how the C++ header symlinks are created in Bazel.

We were hitting this issue very consistently and `7.1.0rc1` appears to have fixed the issue for us.

**Comment by Gormo** [comment]
@freeformstu, The issue still occurs frequently on 7.1.0 when running with bazel modules enabled.

**Comment by fmeum** [comment]
Has anyone been able to reproduce this issue locally with Bazel 7.1.0? This issue is still pretty mysterious to me and a local reproducer, even if not deterministically triggering the bug, would be very helpful.

**Comment by chrisabbott** [comment]
> Has anyone been able to reproduce this issue locally with Bazel 7.1.0 and `--noincompatible_sandbox_hermetic_tmp`? This issue is still pretty mysterious to me and a local reproducer, even if not deterministically triggering the bug, would be very helpful.

@fmeum I was able to reproduce the same issue under Bazel `7.1.0` with `rules_foreign_cc` version `0.10.1`.

I had to add _both_ `--noincompatible_sandbox_hermetic_tmp` and `--noexperimental_merged_skyframe_analysis_execution` to resolve this.

**Comment by fmeum** [comment]
@chrisabbott Could you share an example with which you observe this behavior?

**Comment by chrisabbott** [comment]
> @chrisabbott Could you share an example with which you observe this behavior?

@fmeum Yep! Here you go:

**`.bazelversion`**
```
7.1.0
```

**`MODULE.bazel`**
```
module(
    name = "example",
    version = "0.0",
)

http_archive = use_repo_rule("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")

bazel_dep(
    name = "rules_foreign_cc",
    version = "0.10.1"
)

_ALL_CONTENT = """\
filegroup(
    name = "all_srcs",
    srcs = glob(["**"]),
    visibility = ["//visibility:public"],
)
"""

# Add spdlog 1.13.0 as a third-party dependency.
SPDLOG_VERSION = "1.13.0"
SPDLOG_INTEGRITY = "sha256-n2dju3b/99s3H1czYmyDNS7dfFeJlQGrACSPr62cxQQ="

http_archive(
    name = "spdlog",
    build_file_content = _ALL_CONTENT,
    strip_prefix = "spdlog-{}".format(SPDLOG_VERSION),
    urls = [
        "https://github.com/gabime/spdlog/archive/v{}.zip".format(SPDLOG_VERSION),
    ],
    integrity = SPDLOG_INTEGRITY,
)
```

**`third_party/BUILD.bazel`**

```
load("@rules_foreign_cc//foreign_cc:defs.bzl", "cmake")

cmake(
    name = "spdlog",
    lib_source = "@spdlog//:all_srcs",
    out_static_libs = ["libspdlog.a"],
)
```

**Comment by fmeum** [comment]
@chrisabbott Sorry, I didn't see your edit before: If that flag fixes the issue, I'm pretty sure it's the deterministic failure described in https://github.com/bazelbuild/bazel/issues/21215.

**Comment by Gormo** [comment]
@fmeum, we have --noincompatible_sandbox_hermetic_tmp enabled on 7.1.0, but we still see the error frequently in CI when bazel modules is enabled. (It's actually as frequent so it's stopping us from migrating to bazel modules). It's undeterministic and difficult to understand what triggers it but if you have proposals for relevant log points I can enable those, or even patch bazel for a test if needed.

**Comment by fmeum** [comment]
@Gormo Could you try to bisect this down to a smaller commit range by setting `USE_BAZEL_VERSION` to a commit between 6.4.0 and 7.0.0 with Bazelisk? That could help us understand which kind of change caused this regression.

**Comment by linzhp** [comment]
One thing that we noticed in the past few months after upgrading to Bazel 7 was files in Bazel's output_base mysteriously missing from CI machines, leading to errors like:

```
ERROR: no such package '@@com_github_aws_aws_sdk_go//service/cloudfront/sign': BUILD file not found in directory 'service/cloudfront/sign' of external repository @@com_github_aws_aws_sdk_go. Add a BUILD file to a directory to mark it as a package.
```

or

```
failed to fetch com_github_bazelbuild_remote_apis: fetch_repo: /<truncated>/external/go_sdk/bin/go /<truncated>/external/go_sdk/bin/go mod download -json -modcacherw github.com/bazelbuild/remote-apis@v0.0.0-20231207155409-7e33c12ee961: fork/exec /<truncated>/external/go_sdk/bin/go: no such file or directory
```

This kind of error can happen to any external repo. The error messagess above are just examples. We do have a process on CI machine to clean cache directories, including Bazel output_base when the disk is close to full. However, the cleaning process is to rotate the whole cache directory, not deleting files individually. So it shouldn't cause partial deletion of output_base.


When this error happens, all subsequent builds from that CI machine would fail with similar errors until we run `bazel clean --expunge` there.

I can imaging partial deletion of output_base caused dangling symbolic link discussed on this ticket.

**Comment by fmeum** [comment]
The issues described in this thread that aren't fixed by `--noincompatible_sandbox_hermetic_tmp` are consistent with https://github.com/bazelbuild/bazel/issues/22073: Any kind of symlink pointing into an external repo's directory under the execroot could be dangling due to the race in that issue. I would thus recommend everyone in this thread to try out a fix when it lands.

I don't know what could be causing the issue @linzhp described though.

**Comment by rdesgroppes** [comment]
I faced the issue with `jansson`:
1. `deps.bzl`:
```py
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")

def deps():
    ref = "2.14"
    http_archive(
        name = "jansson",
        build_file = "//my:jansson.BUILD",
        sha256 = "5798d010e41cf8d76b66236cfb2f2543c8d082181d16bc3085ab49538d4b9929",
        strip_prefix = "jansson-{}".format(ref),
        url = "https://github.com/akheron/jansson/releases/download/v{}/jansson-{}.tar.gz".format(ref, ref),
    )
```
2. `jansson.BUILD`:
```py
load("@rules_foreign_cc//foreign_cc:defs.bzl", "cmake")
filegroup(name = "lib_source", srcs = glob(["**"]))
cmake(name = "libjansson", lib_source = ":lib_source", visibility = ["//visibility:public"])
```
> ERROR: /my/external/jansson/BUILD.bazel:9:6: Error while validating output TreeArtifact File:[[<execution_root>]bazel-out/k8-fastbuild/bin]external/jansson/libjansson/include : Child jansson.h of tree artifact /my/execroot/__main__/bazel-out/k8-fastbuild/bin/external/jansson/libjansson/include is a dangling symbolic link
> ERROR: /my/external/jansson/BUILD.bazel:9:6: Foreign Cc - CMake: Building libjansson failed: not all outputs were created or valid
> Target @@jansson//:libjansson failed to build

Workaround:
```diff
         build_file = "//my:jansson.BUILD",
+        patches = ["//my:jansson.patch"],  # https://github.com/bazelbuild/bazel/issues/20886
```
3. patch
```patch
--- CMakeLists.txt
+++ CMakeLists.txt
@@ -271,2 +271,3 @@
-file (COPY ${CMAKE_CURRENT_SOURCE_DIR}/src/jansson.h
-           DESTINATION ${CMAKE_CURRENT_BINARY_DIR}/include/)
+configure_file (${CMAKE_CURRENT_SOURCE_DIR}/src/jansson.h
+                ${CMAKE_CURRENT_BINARY_DIR}/include/jansson.h
+                COPYONLY)
@@ -298 +299 @@
-   ${CMAKE_CURRENT_SOURCE_DIR}/src/jansson.h)
+   ${CMAKE_CURRENT_BINARY_DIR}/include/jansson.h)
```

**Comment by fmeum** [comment]
Since it's pretty likely that this is fixed by https://github.com/bazelbuild/bazel/commit/52adf0b3d14d4af60ece33c95b94be058bafb41b, I will close this issue for now. 

If you can still reproduce your issue with a version of Bazel including this commit (currently `last_green`, but the fix is in the process of being merged into the 7.2.0 branch, which also has all its commits available with `USE_BAZEL_VERSION` after a short wait), then:
* if it's fixed by `--noincompatible_sandbox_hermetic_tmp`, then subscribe to #21215 instead.
* if it persists even with that flag disabled, please file a new issue.
